addons/survey/static/src/js/survey_form.js

```diff
@@ -74,70 +70,235 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend({
+    /**
+    * Follow the submit and handle the transition from one screen to another
+    * Also handle server side validation and displays eventual error messages.
+    */
+    _onSubmitDone: function (result, $target) {
         var self = this;
+
         if (result && !result.error) {
-            window.location = result;
+            this.$(".o_survey_form_content").empty();
+            this.$(".o_survey_form_content").html(result);
+            this.$('div.o_survey_form_date').each(function () {
+                self._initDateTimePicker($(this));
+            });
+            if (this.options.isStartScreen) {
+                this._initTimer();
+                this.options.isStartScreen = false;
+            }
+            this._updateBreadcrumb();
+            if ($target.val() === 'finish') {
+                this._initResultWidget();
+            }
+            this.$('.o_survey_form_content').fadeIn(this.fadeInOutDelay);
+            $("html, body").animate({ scrollTop: 0 }, this.fadeInOutDelay);
         }
         else if (result && result.fields && result.error === 'validation') {
-            var fieldKeys = _.keys(result.fields);
-            _.each(fieldKeys, function (key) {
-                self.$("#" + key + '>.o_survey_question_error').append($('<p>', {text: result.fields[key]})).toggleClass('d-none', false);
-                if (fieldKeys[fieldKeys.length - 1] === key) {
-                    self._scrollToError(self.$('.o_survey_question_error:visible:first').closest('.js_question-wrapper'));
-                }
-            });
-            return false;
+            this.$('.o_survey_form_content').fadeIn(0);
+            this._showErrors(result.fields);
+        } else {
+            var $errorTarget = this.$('.o_survey_error');
+            $errorTarget.removeClass("d-none");
+            this._scrollToError($errorTarget);
         }
-        else {
-            var $target = self.$('.o_survey_error');
-            $target.toggleClass('d-none', false);
-            self._scrollToError($target);
+    },
+
+    // VALIDATION TOOLS
+    // -------------------------------------------------------------------------
+    /**
+    * Validation is done in frontend before submit to avoid latency from the server.
+    * If the validation is incorrect, the errors are displayed before submitting and
+    * fade in / out of submit is avoided.
+    *
+    * Each question type gets its own validation process.
+    *
+    * There is a special use case for the 'required' questions, where we use the constraint
+    * error message that comes from the question configuration ('constr_error_msg' field).
+    *
+    * @private
+    */
+    _validateForm: function ($form, formData) {
+        var self = this;
+        var errors = {};
+        var validationEmailMsg = _t("This answer must be an email address.");
+        var validationDateMsg = _t("This is not a date");
+
+        this._resetErrors();
+
+        var data = {};
```

Comment:
Is data really necessary ?

Reply:
I'm not sure, it's easier for some checks like for "choice" questions.

Do you have something else in mind?